Zend/Optimizer/zend_optimizer.c

```diff
@@ -891,16 +891,23 @@ zend_function *zend_optimizer_get_called_func(
 					&op_array->scope->function_table, method_name);
 				if (fbc) {
 					bool is_private = (fbc->common.fn_flags & ZEND_ACC_PRIVATE) != 0;
-					bool is_final = (fbc->common.fn_flags & ZEND_ACC_FINAL) != 0;
-					bool same_scope = fbc->common.scope == op_array->scope;
 					if (is_private) {
 						/* Only use private method if in the same scope. We can't even use it
 						 * as a prototype, as it may be overridden with changed signature. */
+						bool same_scope = fbc->common.scope == op_array->scope;
 						return same_scope ? fbc : NULL;
 					}
 					/* If the method is non-final, it may be overridden,
 					 * but only with a compatible method signature. */
-					*is_prototype = !is_final;
+					/* We know it isn't a prototype if it can be overridden.
+					 * A method cannot be overridden if the class or method is final, or
+					 * if it's an anonymous class. */
+					bool can_be_overridden = (fbc->common.fn_flags & ZEND_ACC_FINAL) == 0 &&
+											 (fbc->common.scope->ce_flags & ZEND_ACC_FINAL) == 0 &&
+											 (fbc->common.scope->ce_flags & ZEND_ACC_ANON_CLASS) == 0;
+					if (can_be_overridden) {
+						*is_prototype = true;
```

Comment:
It's not a prototype if it can be overridden, right? Also, the `zend_optimizer_get_called_func()` function seems to be less safe (less intuitive to use) for me if it only writes the `is_prototype` when the respective method can be overridden.

Reply:
An initial false value is written at the top of the function body. The reason I used an if condition here is to avoid needless memory writes. But perhaps that looks a little odd indeed.

Reply:
Ah, ok, if it's initialized above then my concern is addressed :)